src/playback/view/playbacktoolbarmodel.cpp

```diff
@@ -59,119 +58,170 @@ QVariant PlaybackToolBarModel::data(const QModelIndex& index, int role) const
         return QVariant();
     }
 
-    const MenuItem& item = m_items.at(index.row());
+    const QVariantMap& item = items().at(index.row()).toMap();
 
     switch (role) {
-    case HintRole: return item.description;
-    case IconRole: return static_cast<int>(item.iconCode);
-    case CodeRole: return QString::fromStdString(item.code);
-    case CheckedRole: return item.state.checked;
-    case IsAdditionalRole: return isAdditionalAction(item.code);
-    case IsPlaybackSettingsRole: return item.code == PLAYBACK_SETTINGS_KEY;
+    case CodeRole: return item["code"];
+    case HintRole: return item["description"];
+    case IconRole: return item["icon"];
+    case CheckedRole: return item["checked"];
+    case SubitemsRole: return item["subitems"];
     }
 
     return QVariant();
 }
 
 int PlaybackToolBarModel::rowCount(const QModelIndex&) const
 {
-    return m_items.count();
+    return items().count();
 }
 
-QHash<int,QByteArray> PlaybackToolBarModel::roleNames() const
+QHash<int, QByteArray> PlaybackToolBarModel::roleNames() const
 {
     static const QHash<int, QByteArray> roles {
         { CodeRole, "code" },
         { HintRole, "hint" },
         { IconRole, "icon" },
         { CheckedRole, "checked" },
-        { IsAdditionalRole, "isAdditional" },
-        { IsPlaybackSettingsRole, "isPlaybackSettings" }
+        { SubitemsRole, "subitems" },
     };
 
     return roles;
 }
 
 void PlaybackToolBarModel::load()
 {
-    beginResetModel();
-    m_items.clear();
-
-    UiActionList additionalActions;
-
-    for (const UiAction& action : currentWorkspaceActions()) {
-        if (isAdditionalAction(action.code)) {
-            additionalActions.push_back(action);
-            continue;
-        }
-
-        m_items << action;
-    }
-
-    m_items << settingsItem();
-
-    for (const UiAction& action : additionalActions) {
-        m_items << action;
-    }
-
-    endResetModel();
-
-    updateState();
+    updateActions();
+    listenActionsStateChanges();
     setupConnections();
 }
 
 void PlaybackToolBarModel::setupConnections()
 {
+    connect(this, &PlaybackToolBarModel::floatingChanged, this, &PlaybackToolBarModel::updateActions);
+
     playbackController()->isPlayAllowedChanged().onNotify(this, [this]() {
-        updateState();
+        emit maxPlayTimeChanged();
         updatePlayTime();
         emit isPlayAllowedChanged();
     });
 
     playbackController()->isPlayingChanged().onNotify(this, [this]() {
-        updateState();
+        onActionsStateChanges({ "play" });
```

Comment:
Better to add a static string at the top of the file for this code because I see it in several places

Reply:
Thanks, fixed!